### ISSUE TITLE
Dumb-Jump using fallback mechanisms

### ISSUE THREAD

**Issue by jojojames** [jacktasia/dumb-jump#172]
Hi, thanks for the awesome package! I'm curious if there was any chance of supporting GoToDefinition in a more
waterfall/fallback driven approach.

I find myself redefining/creating a lot of similar functions so that I can use
dumb-jump as a fallback in case the primary method of GoToDefinition fails.

Along with that, sometimes there are various methods I'd like to try before
giving up, and falling back on dumb-jump. For example, a snippet like this is
common in my own configuration.

```
  (defvar +elisp-navigation-stack '() "Stack used to navigate tags.")

  (defun +elisp-gtd-dwim ()
    "DWIM Goto Definition."
    (interactive)
    (condition-case nil
        (progn
          (call-interactively 'elisp-slime-nav-find-elisp-thing-at-point)
          (push 'elisp-slime-nav-mode +elisp-navigation-stack))
      (error
       (progn
         (call-interactively 'dumb-jump-go)
         (push 'dumb-jump-go +elisp-navigation-stack)))))

  (defun +elisp-gtd-back ()
    "DWIM Pop Goto Definition."
    (interactive)
    (let ((method (pop +elisp-navigation-stack)))
      (cond
       ((eq method 'elisp-slime-nav-mode) (pop-tag-mark))
       ((eq method 'dumb-jump-go) (dumb-jump-back))
       (t
        (message "No tag found to pop.")))))
```

Maybe we can support some kind of list of other commands to be called before we
try using dumb-jump.

The list may contain a 3 tuple list:

1. Whether or not the backend should be called at the time this command is invoked.

This is probably a good one to check for obvious failure conditions like a
language server being down, etc.

2. The actual function to be called. It might be something we'd pass in to
(call-interactively #'symbol) for example.

3. A way to check if the call completed successfully or not. This seems like the
trickiest part. Maybe there's a list of symbols that check for different heuristics.

A few that come to mind are: Error, Mark Moved, Message Buffer changed.

If Error, it's probably a simple try/catch block like above. This may be tricky only
because there might be an asynchrounous component to checking for such a change.

A resulting list could be something along the lines of.

'(((lambda ()
  (if (server-is-up-p)
      t
    nil)) find-definition-using-server mark-moved)
  (t xref-find-definition error))

**Comment by jacktasia** [comment]
Thanks for opening this issue and the kind words. I think this is a really good idea, and I am certainly open to doing something like this, but my initial gut feeling is that what you're describing should be its own, higher level, package (maybe called `smart-jump`).

What do you think about this being its own package? I am going to think about this a bit more and look into how spacemacs does this because I think they do something similar with their layers, but it might not be as smart as what you're describing. Thanks again!

**Comment by jojojames** [comment]
Hmnn, yeah another package would work too. smart-jump would be a good name for it for sure..

**Comment by jacktasia** [comment]
Closing this for now but I am open to discussing more and/or collobrating if you want to take the lead on a `smart-jump` package.

**Comment by jojojames** [comment]
@jacktasia I've written an initial implementation of it here.

https://github.com/jojojames/smart-jump

Would love your input/feedback if you had some free time.

I'll dogfood it in the meantime.

**Comment by jacktasia** [comment]
Awesome! This looks good. I can't comment too much on the implementation without tests, but everything certainly looks sane to me. I know you said this is your initial implementation so sorry if any of this is feedback obvious to you.

* I think your README looks great. I made a small PR with some small tweaks.

* I'd rename `dumb-jump-fallback` to `smart-jump-dumb-fallback` or something that keeps your namespace. Mainly in case this same var gets added to dumb-jump in a different context down the line by someone else and I forget and these start to shadow one another.

* I'd add tests and a CI (circle/travis). This makes it much easier to make big changes down the line and accept PRs.

* I'd submit this to the emacs reddit for feedback and then I'd make a PR to MELPA to add the smart-jump recipe.

I hope this helps. I am happy to expand on anything or comment on anything specific just let me know, but honestly, I don't know much about the emacs `async` library or these smarter jump to definition tools...since I made dumb-jump so I wouldn't have to worry about configuring them :-)
